src/Development/IDE/Core/Shake.hs

```diff
@@ -536,7 +536,7 @@ usesWithStale key files = do
 
 withProgress :: (Eq a, Hashable a) => Var (HMap.HashMap a Int) -> a -> Action b -> Action b
 withProgress var file = actionBracket (f succ) (const $ f pred) . const
-    where f shift = modifyVar_ var $ return . HMap.alter (Just . shift . fromMaybe 0) file
+    where f shift = modifyVar_ var $ \x -> return (HMap.alter (Just . shift . fromMaybe 0) file x)
```

Comment:
The linked blog post by @nomeata explains why this can lead to performance gains, but it isn't clear that it fixes a space leak

Reply:
I profiled a session where I repeatedly modified the file and could observe this function closure growing in the profile. I eta-expanded and the band was no longer present in the profile. So I agree it's not clear why this fixes the leak I don't think it should block merging.

Reply:
Can we make it `evaluate (HMap.alter` - I suspect the space leak is removed because GHC manages to optimise a little bit more in some place. The `evaluate` would mean the operation guarantees to run in the IO monad and not stack up.

Reply:
:+1: